### PR TITLE
Compile time env available from cython command line options

### LINKED ISSUE #2314: compile_time_env on command line
It would be nice to be able to defile "compile time env" variable in when calling cython from the command line.

I will summit a PR once I got something working

### PR DESCRIPTION
**Pull request by kif**
I used the "-E" short option but if you believe other letters are better suited, I can change.
close #2314

### CODE REVIEW

File: Cython/Compiler/Options.py [hunk 1]
```diff
@@ -441,3 +441,73 @@ def parse_directive_list(s, relaxed_bool=False, ignore_unknown=False,
...
+
+def parse_compile_time_env(s, current_settings=None):
+    """
+    Parses a comma-separated list of pragma options. Whitespace
+    is not considered.
+
+    >>> parse_compile_time_env('      ')
+    {}
+    >>> (parse_compile_time_env('boundscheck=True') ==
+    ... {'boundscheck': True})
```
Comment (scoder) [inline Cython/Compiler/Options.py]: This is not a good example, because it suggests directive names to be evaluated. The compile time env has nothing to do with directives.

File: Cython/Compiler/Options.py [hunk 2]
```diff
@@ -441,3 +441,73 @@ def parse_directive_list(s, relaxed_bool=False, ignore_unknown=False,
...
+        result = {}
+    else:
+        result = current_settings
+    for item in s.split(','):
+        item = item.strip()
+        if not item:
+            continue
+        if '=' not in item:
+            raise ValueError('Expected "=" in option "%s"' % item)
+        name, value = [s.strip() for s in item.strip().split('=', 1)]
```
Comment (scoder) [inline Cython/Compiler/Options.py]: No need for stripping the original string here, since we need to strip each part anyway.

File: Cython/Compiler/Options.py [hunk 3]
```diff
@@ -423,7 +423,7 @@ def parse_directive_list(s, relaxed_bool=False, ignore_unknown=False,
         item = item.strip()
         if not item:
             continue
-        if not '=' in item:
+        if '=' not in item:
```
Comment (scoder) [inline Cython/Compiler/Options.py]: It's a bit sad that these changes slipped in. Probably a commit hook on your side or so.
While I think that cleaning them up is a good thing, it should better not happen as part of an unrelated commit.

File: Cython/Build/Cythonize.py [hunk 4]
```diff
@@ -139,6 +139,9 @@ def parse_args(args):
     parser.add_option('-X', '--directive', metavar='NAME=VALUE,...', dest='directives',
                       type=str, action='callback', callback=parse_directives, default={},
                       help='set a compiler directive')
+    parser.add_option('-E', '--compile-time-env', metavar='NAME=VALUE,...', dest='compile_time_env',
+                      type=str, action='callback', callback=Options.parse_compile_time_env, default={},
+                      help='set a compile time environment variables')
```
Comment (scoder) [inline Cython/Build/Cythonize.py]: "a ... variable"

File: Cython/Build/Cythonize.py [hunk 5]
```diff
@@ -139,6 +139,9 @@ def parse_args(args):
     parser.add_option('-X', '--directive', metavar='NAME=VALUE,...', dest='directives',
                       type=str, action='callback', callback=parse_directives, default={},
                       help='set a compiler directive')
+    parser.add_option('-E', '--compile-time-env', metavar='NAME=VALUE,...', dest='compile_time_env',
+                      type=str, action='callback', callback=Options.parse_compile_time_env, default={},
```
Comment (scoder) [inline Cython/Build/Cythonize.py]: I can see in the surrounding code where this is coming from, but it seems wrong to say `type=str` when the actual value is a dict. And I'm also not sure that this allows passing `-E` multiple times.

**Comment by scoder** [comment]
I'm a bit torn whether `-E` is better than `-D`, but I'm not entirely sure that we'll never want to use `-D` to pass macro definitions down to the C compiler in `cythonize`, so ...

**Comment by kif** [comment]
On Fri, 08 Jun 2018 23:14:01 -0700
scoder <notifications@github.com> wrote:

> I'm a bit torn whether `-E` is better than `-D`, but I'm not entirely sure that we'll never want to use `-D` to pass macro definitions down to the C compiler in `cythonize`, so ...

unfortunately -D is already used for no doc-strings ... the alternative
I see is -Y as -X is for compiler directives...

I'll fix the docstrings for tests to make them more explicit.

Concerning the 'not "=" in', pep8 recommands to replace it with its
direct equivalent. Having a specific PR for this tend to reduce the
agility of the project. but as you want, you are the reviewer :)

Cheers

Jerome

**Comment by scoder** [comment]
The spacing/PEP8 changes don't need to be a separate PR, really, but they should not be part of an unrelated commit. Having them as a separate commit would be ok. (I admit that I'm guilty of doing such cleanups myself from time to time, and not always taking the time to separate them out either. We're all human...)

**Comment by scoder** [comment]
What I'm saying is: my other comments are more important than the one about the style changes. :)

File: Cython/Compiler/Options.py (hunk 1 above)
Comment (kif) [reply Cython/Compiler/Options.py]: I agree, I'll fix that with dummy variable names

File: Cython/Build/Cythonize.py (hunk 5 above)
Comment (kif) [reply Cython/Build/Cythonize.py]: The type is the input type, before the action of the callback.
Setting it to "dict" would require the definition of what is a dict in OptParser, how to parse it, ... and duplicate most of the code already present. 
When not defining it, the callback gets None instead of "" and most of them fail in parsing the string.

I am not very familiar with cythonize in the command line but apparently it works now as well.

File: Cython/Compiler/Options.py (hunk 1 above)
Comment (kif) [reply Cython/Compiler/Options.py]: I updated the doctest to be more realistic and less misleading for other developers.
